### PR TITLE
VAE training sample script

### PR DESCRIPTION
**Pull request by Pie31415**
PR for Issue #3726

Todos
- [x] implement training loop for VAE
- [x] KL loss implementation
- [ ] evaluate performance of VAE training
- [ ] fix script to work for mixed precision
- [ ] integration with a1111

### CODE REVIEW

**Comment by Pie31415** [comment]
[06/24/2023] VAE fine-tuning runs successfully but will need to test/evaluate results of images.

**Comment by zhuliyi0** [comment]
I tried to run it on a 4070 with 12G vram, and it gave OOM. I then ran it on a colab T4, the first epoch goes normally using only 8.5G vram. Strange that 4070 wants more vram.

On colab, the validation after first epoch gave error:

2023-07-08 01:39:22.493265: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT
07/08/2023 01:39:24 - INFO - __main__ - Distributed environment: NO
Num processes: 1
Process index: 0
Local process index: 0
Device: cuda

Mixed precision type: no

Resolving data files: 100% 38/38 [00:00<00:00, 31542.36it/s]
07/08/2023 01:39:25 - WARNING - datasets.builder - Found cached dataset imagefolder (/root/.cache/huggingface/datasets/imagefolder/lvchengguiyuxi_try11_highriseOnly-1a6b424a6b99295d/0.0.0/37fbb85cc714a338bea574ac6c7d0b5be5aff46c1862c1989b20e0771199e93f)
100% 1/1 [00:00<00:00, 441.65it/s]
07/08/2023 01:39:33 - INFO - __main__ - ***** Running training *****
07/08/2023 01:39:33 - INFO - __main__ -   Num examples = 33
07/08/2023 01:39:33 - INFO - __main__ -   Num test samples = 5
07/08/2023 01:39:33 - INFO - __main__ -   Num Epochs = 100
07/08/2023 01:39:33 - INFO - __main__ -   Instantaneous batch size per device = 1
07/08/2023 01:39:33 - INFO - __main__ -   Total train batch size (w. parallel, distributed & accumulation) = 1
07/08/2023 01:39:33 - INFO - __main__ -   Gradient Accumulation steps = 1
Steps:   1% 33/3300 [02:37<4:07:24,  4.54s/it, lr=0.0001, step_loss=0.00202]07/08/2023 01:42:11 - INFO - __main__ - Running validation... 
╭───────────────────── Traceback (most recent call last) ──────────────────────╮
│ /content/vae_training.py:518 in <module>                                     │
│                                                                              │
│   515                                                                        │
│   516                                                                        │
│   517 if __name__ == "__main__":                                             │
│ ❱ 518 │   main()                                                             │
│   519                                                                        │
│                                                                              │
│ /content/vae_training.py:506 in main                                         │
│                                                                              │
│   503 │   │                                                                  │
│   504 │   │   if accelerator.is_main_process:                                │
│   505 │   │   │   if epoch % args.validation_epochs == 0:                    │
│ ❱ 506 │   │   │   │   log_validation(test_dataloader, vae, accelerator, weig │
│   507 │                                                                      │
│   508 │   # Create the pipeline using the trained modules and save it.       │
│   509 │   accelerator.wait_for_everyone()                                    │
│                                                                              │
│ /content/vae_training.py:65 in log_validation                                │
│                                                                              │
│    62 │                                                                      │
│    63 │   for tracker in accelerator.trackers:                               │
│    64 │   │   if tracker.name == "tensorboard":                              │
│ ❱  65 │   │   │   np_images = np.stack([np.asarray(img) for img in images])  │
│    66 │   │   │   tracker.writer.add_images("Original (left) / Reconstructio │
│    67 │   │   elif tracker.name == "wandb":                                  │
│    68 │   │   │   tracker.log(                                               │
│                                                                              │
│ /content/vae_training.py:65 in <listcomp>                                    │
│                                                                              │
│    62 │                                                                      │
│    63 │   for tracker in accelerator.trackers:                               │
│    64 │   │   if tracker.name == "tensorboard":                              │
│ ❱  65 │   │   │   np_images = np.stack([np.asarray(img) for img in images])  │
│    66 │   │   │   tracker.writer.add_images("Original (left) / Reconstructio │
│    67 │   │   elif tracker.name == "wandb":                                  │
│    68 │   │   │   tracker.log(                                               │
│                                                                              │
│ /usr/local/lib/python3.10/dist-packages/torch/_tensor.py:970 in __array__    │
│                                                                              │
│    967 │   │   if has_torch_function_unary(self):                            │
│    968 │   │   │   return handle_torch_function(Tensor.__array__, (self,), s │
│    969 │   │   if dtype is None:                                             │
│ ❱  970 │   │   │   return self.numpy()                                       │
│    971 │   │   else:                                                         │
│    972 │   │   │   return self.numpy().astype(dtype, copy=False)             │
│    973                                                                       │
╰──────────────────────────────────────────────────────────────────────────────╯
RuntimeError: Can't call numpy() on Tensor that requires grad. Use 
tensor.detach().numpy() instead.
Steps:   1% 33/3300 [02:43<4:29:40,  4.95s/it, lr=0.0001, step_loss=0.00202]

File: examples/vae/README.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+The command to train a VAE model on a custom dataset (`DATASET_NAME`):
+
+```bash
+python train_vae.py --mixed_precision="no" \
+    --pretrained_model_name_or_path="stabilityai/stable-diffusion-2-1" \
+    --dataset_name="<DATASET_NAME>" \
+    --train_batch_size=1 \
+    --gradient_accumulation_steps=4 \
+    --gradient_checkpointing
```
Comment (epitaque) [inline examples/vae/README.md:33]:
```suggestion
    --gradient_checkpointing \
```

File: examples/vae/train_vae.py [hunk 2]
```diff
@@ -0,0 +1,546 @@
...
+        "--checkpoints_total_limit",
+        type=int,
+        default=None,
+        help=(
+            "Max number of checkpoints to store. Passed as `total_limit` to the `Accelerator` `ProjectConfiguration`."
+            " See Accelerator::save_state https://huggingface.co/docs/accelerate/package_reference/accelerator#accelerate.Accelerator.save_state"
+            " for more docs"
+        ),
+    )
+    parser.add_argument(
```
Comment (zhuliyi0) [inline examples/vae/train_vae.py:246]: add test_data_dir:

`    parser.add_argument(
        "--test_data_dir",
        type=str,
        default=None,
        help=(
            "If not None, will override test_samples arg and use data inside this dir as test dataset."
        ),
    )`

File: examples/vae/train_vae.py [hunk 3]
```diff
@@ -0,0 +1,546 @@
...
+        reconstructions = vae_model(x).sample
+        images.append(
+            torch.cat([sample["pixel_values"].cpu(), reconstructions.cpu()], axis=0)
+        )
+
+    for tracker in accelerator.trackers:
+        if tracker.name == "tensorboard":
+            np_images = np.stack([np.asarray(img) for img in images])
+            tracker.writer.add_images(
+                "Original (left) / Reconstruction (right)", np_images, epoch
```
Comment (zhuliyi0) [inline examples/vae/train_vae.py:61]: change file name to be compatible with Windows:

`"Original (left)-Reconstruction (right)", np_images, epoch`

File: examples/vae/train_vae.py [hunk 4]
```diff
@@ -0,0 +1,546 @@
...
+            transforms.Normalize([0.5], [0.5]),
+        ]
+    )
+
+    def preprocess(examples):
+        images = [image.convert("RGB") for image in examples[image_column]]
+        examples["pixel_values"] = [train_transforms(image) for image in images]
+        return examples
+
+    with accelerator.main_process_first():
```
Comment (zhuliyi0) [inline examples/vae/train_vae.py:390]: support load testset from test_data_set folder:

`    with accelerator.main_process_first():
        # Load test data from test_data_dir
        if(args.test_data_dir is not None and args.train_data_dir is not None):
            logger.info(f"load test data from {args.test_data_dir}")
            test_dir = os.path.join(args.test_data_dir, "**")        
            test_dataset = load_dataset(
                "imagefolder",
                data_files=test_dir,
                cache_dir=args.cache_dir,
            )
            # Set the training transforms
            train_dataset = dataset["train"].with_transform(preprocess)
            test_dataset = test_dataset["train"].with_transform(preprocess)
        # Split into train/test
        else:
            dataset = dataset["train"].train_test_split(test_size=args.test_samples)        
            # Set the training transforms
            train_dataset = dataset["train"].with_transform(preprocess)
            test_dataset = dataset["test"].with_transform(preprocess)`

File: examples/vae/train_vae.py [hunk 5]
```diff
@@ -0,0 +1,546 @@
...
+
+
+def log_validation(test_dataloader, vae, accelerator, weight_dtype, epoch):
+    logger.info("Running validation... ")
+
+    vae_model = accelerator.unwrap_model(vae)
+    images = []
+    for _, sample in enumerate(test_dataloader):
+        x = sample["pixel_values"].to(weight_dtype)
+        reconstructions = vae_model(x).sample
```
Comment (zhuliyi0) [inline examples/vae/train_vae.py:52]: this line seems to cause VRAM leak on Windows system.

File: examples/vae/train_vae.py [hunk 6]
```diff
@@ -0,0 +1,546 @@
...
+        images = [image.convert("RGB") for image in examples[image_column]]
+        examples["pixel_values"] = [train_transforms(image) for image in images]
+        return examples
+
+    with accelerator.main_process_first():
+        # Split into train/test
+        dataset = dataset["train"].train_test_split(test_size=args.test_samples)
+        # Set the training transforms
+        train_dataset = dataset["train"].with_transform(preprocess)
+        test_dataset = dataset["test"].with_transform(preprocess)
```
Comment (zhuliyi0) [inline examples/vae/train_vae.py:395]: support load testset from test_data_set folder

```suggestion
    with accelerator.main_process_first():
        # Load test data from test_data_dir
        if(args.test_data_dir is not None and args.train_data_dir is not None):
            logger.info(f"load test data from {args.test_data_dir}")
            test_dir = os.path.join(args.test_data_dir, "**")        
            test_dataset = load_dataset(
                "imagefolder",
                data_files=test_dir,
                cache_dir=args.cache_dir,
            )
            # Set the training transforms
            train_dataset = dataset["train"].with_transform(preprocess)
            test_dataset = test_dataset["train"].with_transform(preprocess)
        # Split into train/test
        else:
            dataset = dataset["train"].train_test_split(test_size=args.test_samples)        
            # Set the training transforms
            train_dataset = dataset["train"].with_transform(preprocess)
            test_dataset = dataset["test"].with_transform(preprocess)
```

File: examples/vae/train_vae.py [hunk 7]
```diff
@@ -0,0 +1,546 @@
...
+            " See Accelerator::save_state https://huggingface.co/docs/accelerate/package_reference/accelerator#accelerate.Accelerator.save_state"
+            " for more docs"
+        ),
+    )
+    parser.add_argument(
+        "--test_samples",
+        type=int,
+        default=4,
+        help="Number of images to remove from training set to be used as validation.",
+    )
```
Comment (zhuliyi0) [inline examples/vae/train_vae.py:251]: add new argument test_data_dir, for dedicated test data folder
```suggestion
    )
   parser.add_argument(
        "--test_data_dir",
        type=str,
        default=None,
        help=(
            "If not None, will override test_samples arg and use data inside this dir as test dataset."
        ),
    )
```

File: examples/vae/train_vae.py (hunk 3 above)
Comment (zhuliyi0) [inline examples/vae/train_vae.py:61]: change file name to be compatible with Windows
```suggestion           
                "Original (left)-Reconstruction (right)", np_images, epoch
```

File: examples/vae/train_vae.py (hunk 5 above)
Comment (epitaque) [reply examples/vae/train_vae.py:52]: can confirm, adding `@torch.no_grad()` before `log_validation` fixed my CUDA out of memory errors.

Comment (zhuliyi0) [reply examples/vae/train_vae.py:52]: Yes mine is fixed too. Thanks!

File: examples/vae/train_vae.py [hunk 8]
```diff
@@ -0,0 +1,546 @@
...
+                "mse": mse_loss.detach().item(),
+                "lpips": lpips_loss.detach().item(),
+                "kl": kl_loss.detach().item(),
+            }
+            accelerator.log(logs)
+            progress_bar.set_postfix(**logs)
+
+        if accelerator.is_main_process:
+            if epoch % args.validation_epochs == 0:
+                log_validation(test_dataloader, vae, accelerator, weight_dtype, epoch)
```
Comment (zhuliyi0) [inline examples/vae/train_vae.py:534]:
```suggestion
         with torch.no_grad():
                log_validation(test_dataloader, vae, accelerator, weight_dtype, epoch)
```

this fixes the VRAM issue as suggested by @epitaque

File: examples/vae/train_vae.py [hunk 9]
```diff
@@ -0,0 +1,546 @@
...
+    first_epoch = 0
+
+    progress_bar = tqdm(
+        range(global_step, args.max_train_steps),
+        disable=not accelerator.is_local_main_process,
+    )
+    progress_bar.set_description("Steps")
+
+    lpips_loss_fn = lpips.LPIPS(net="alex").to(accelerator.device)
+
```
Comment (zhuliyi0) [inline examples/vae/train_vae.py:477]:
```suggestion

    #initial validation as baseline
    with torch.no_grad():
        log_validation(test_dataloader, vae, accelerator, weight_dtype, 0)
```

one validation before training start as baseline for comparison.

File: examples/vae/train_vae.py [hunk 10]
```diff
@@ -0,0 +1,548 @@
...
+            with accelerator.accumulate(vae):
+                target = batch["pixel_values"].to(weight_dtype)
+
+                # https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/autoencoder_kl.py
+                posterior = vae.encode(target).latent_dist
+                z = posterior.mode()
+                pred = vae.decode(z).sample
+
+                kl_loss = posterior.kl().mean()
+                mse_loss = F.mse_loss(pred, target, reduction="mean")
```
Comment (JamesHao-ml) [inline examples/vae/train_vae.py:492]: In original stable-diffusion repo and SDXL repo, the vae loss is averaged over batch dim, which means they are summed in channel*height*width dims. Is this the correct way to average reconstruction loss?
https://github.com/CompVis/stable-diffusion/blob/21f890f9da3cfbeaba8e2ac3c425ee9e998d5229/ldm/modules/losses/contperceptual.py#L58

**Review by JunzheJosephZhu — COMMENTED** [review]
lgtm

**Comment by lavinal712** [comment]
Is there any progress now?

**Pull request closed without merging**